R/MsBackendMemory.R

```diff
@@ -0,0 +1,657 @@
+    object
+})
+
+#' @rdname hidden_aliases
+setMethod("scanIndex", "MsBackendMemory", function(object) {
+    .get_column(object@spectraData, "scanIndex")
+})
+
+#' @rdname hidden_aliases
+setMethod("selectSpectraVariables", "MsBackendMemory",
+          function(object, spectraVariables = spectraVariables(object)) {
+              if (!all(spectraVariables %in% spectraVariables(object)))
+                  stop("Spectra variables ",
+                       paste(spectraVariables[!(spectraVariables %in%
+                                                spectraVariables(object))],
+                             collapse = ", "), " not available")
+              ## spectraData
+              keep <- intersect(spectraVariables, colnames(object@spectraData))
+              object@spectraData <- object@spectraData[, keep, drop = FALSE]
+              ## peaksData
+              if (length(object@peaksData)) {
+                  have <- colnames(object@peaksData[[1L]])
+                  keep <- intersect(spectraVariables, have)
+                  if (length(keep) < length(have))
+                      object@peaksData <- lapply(object@peaksData, function(z)
+                          z[, keep, drop = FALSE])
+              }
+              ## peaksDataFrame
+              if (length(object@peaksDataFrame)) {
+                  have <- colnames(object@peaksDataFrame[[1L]])
+                  keep <- intersect(spectraVariables, have)
+                  if (!length(keep)) {
+                      object@peaksDataFrame <- list()
+                  } else {
+                      if (length(keep) < length(have))
+                          object@peaksDataFrame <- lapply(
+                              object@peaksDataFrame, function(z)
+                                  z[, keep, drop = FALSE])
+                  }
+              }
+              msg <- .valid_spectra_data_required_columns(object@spectraData)
+              if (length(msg))
+                  stop(msg)
+              validObject(object)
+              object
+})
+
+#' @rdname hidden_aliases
+setMethod("smoothed", "MsBackendMemory", function(object) {
+    .get_column(object@spectraData, "smoothed")
+})
+
+#' @rdname hidden_aliases
+#'
+#' @aliases smoothed<-,MsBackendMemory-method
+setReplaceMethod("smoothed", "MsBackendMemory", function(object, value) {
+    value_len <- length(value)
+    value_type <- is.logical(value)
+    if (value_type && (value_len == 1L || value_len == length(object)))
+        object@spectraData$smoothed <- value
+    else
+        stop("'value' has to be a 'logical' of length 1 or ", length(object))
+    validObject(object)
+    object
+})
+
+#' @rdname hidden_aliases
+#'
+#' @importFrom methods as
+#'
+#' @importFrom S4Vectors SimpleList
+#'
+#' @importMethodsFrom S4Vectors lapply
+setMethod(
+    "spectraData", "MsBackendMemory",
+    function(object, columns = spectraVariables(object)) {
+        as(.df_spectra_data(object, columns), "DataFrame")
```

Comment:
Does this have to return a `DataFrame`? (is this needed for compatibility reasons?)

Reply:
exactly. That's our convention.

Reply:
Back from the days where we were not yet aware of the performance difference...

Reply:
But do we want to keep this? Wouldn't it make more sense to return the slot as it is?

Reply:
I think it's cleaner that way. We defined in the `MsBackend` specification that the `spectraData` returns a `DataFrame` and that `spectraData<-` expects a `DataFrame`. How the data is stored internally is up to the backend, but we have to be consistent what each backend returns or takes. I think it's important to define a clean and consistent interface. Otherwise we could get into trouble in `Spectra`, needing to check what is returned by the backend, reformatting etc.